### PR TITLE
Adding further options to limit built files and honoring other build options

### PR DESCRIPTION
**Pull request by damienpontifex**
I was using this library as a submodule and linking via make and noticing the samples were built and even though I had disabled MQTT as I was just using AMQP these targets were still getting built.

Added options that default to behaviour before change, but allows these other directories to be skipped. Reduces build times when just building the library.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@damienpontifex__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <p>The agreement was validated by Microsoft and real humans are currently evaluating your PR.</p>

TTYL, AZPRBOT;

**Comment by zolvarga** [comment]
Hi Damien,

Thanks for the contribution to our repository!
Sorry for the delay, finally we got time to take a look at your PR.

There are couple of questions about your pull request:
- Introducing 2 independent options creates a matrix of possible configurations (ON-ON, ON-OFF, OFF-OFF, OFF-ON). We should make sure the invalid configuration are excluded.
- The "build_samples" option should be applied to all the samples in the solution: azure-c-shared-utility, azure-uamqp-c, azure-umqtt-c, iothub_service_client

Could you please consider to update the PR with these changes?
 
Best Regards,
Zoltan

**Comment by damienpontifex** [comment]
Thanks @zolvarga firstly the easy one of the 'build_samples': yes I had thought this, but most of the other repos are submodules, so I would create a separate PR in each of their respective repositories?

I will review validating options for the possible configurations

**Comment by zolvarga** [comment]
Yes, we should propagate the option to the submodules.

**Comment by damienpontifex** [comment]
@zolvarga I think that's all of the sample exclusions for this repo. I will make separate PRs to the other repos if this is merged in so the option gets propagated through.

As the samples and service client are not dependencies (plus the service samples are ignored if service client is ignored), I don't think there are any invalid configurations to check for.

Also added a check to ensure at least one transport provider is selected.

I can rebase master before merging if you're happy to accept the changes.

File: python/build_all/linux/build.sh [hunk 1]
```diff
@@ -31,8 +31,7 @@ process_args()
 process_args $*
 
 # instruct C builder to include python library and to skip tests
-
-./c/build_all/linux/build.sh --build-python $PYTHON_VERSION --skip-unittests $*
+./c/build_all/linux/build.sh --build-python $PYTHON_VERSION --skip-unittests --skip-samples --skip-service-client $*
```
Comment (tameraw) [inline python/build_all/linux/build.sh]: @damienpontifex - Can you please revert this line. We would like to build the samples & service client by default.

**Comment by tameraw** [comment]
@damienpontifex - Thanks for the contribution. I can address the comment myself and merge it in.

File: python/build_all/linux/build.sh (hunk 1 above)
Comment (damienpontifex) [inline python/build_all/linux/build.sh]: @tameraw this is for the python client library though. I would assume if you are running the build script from here, you want to focus on the python outputs only? It also reduces the compile requirements when building on smaller devices i.e. raspberry pi as we only go on to call `make` so all targets that are set are built.

It is still defaults to build samples and service client when running the script from the 'c' build_all folder.

**Pull request closed without merging**